### PR TITLE
Add lease manager worker documentation

### PR DESCRIPTION
**Pull request by nvinuesa**
This patch adds documentation to the lease manager worker.

## Checklist

*If an item is not applicable, use `~strikethrough~`.*

- [ ] ~Code style: imports ordered, good names, simple structure, etc~
- [X] Comments saying why design decisions were made
- [ ] ~Go unit tests, with comments saying what you're testing~
- [ ] ~[Integration tests](https://github.com/juju/juju/tree/main/tests), with comments saying what you're testing~
- [ ] ~[doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages~

## Documentation changes

Adds documentation as `doc.go` comments to the `lease` worker package.

### CODE REVIEW

File: worker/lease/doc.go [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+// Package lease is a worker that manages leases, which are used by individual
+// juju workers.
+// The manager works by using these operations:
+//   - claim
+//   - revoke
+//   - pin
+//   - query
+//   - block
```
Comment (barrettj12) [inline worker/lease/doc.go:11]: Just a thought on structure - would it make more sense to have the descriptions inside the list, rather than listing them all and then describing them in a follow up paragraph? For example
```
// The manager works by using these operations:
//   - claim: Workers will claim a lease, and they are either attributed or blocked. The
// lease manager will keep track of all the blocked claims. ...
//   - revoke: ...
//   - ...
```

File: worker/lease/doc.go [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+//
+// Workers will claim a lease, and they are either attributed or blocked. The
```
Comment (barrettj12) [inline worker/lease/doc.go:13]: what do "attributed" and "blocked" mean?

File: worker/lease/doc.go [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+// lease manager will keep track of all the blocked claims. When a lease
+// expires or gets revoked, then one of the claims of other workers will be
+// unblocked, thus assigning the lease to that worker.
```
Comment (barrettj12) [inline worker/lease/doc.go:16]: How is the unblocked worker decided? Is it random? Is there a queue?

**Review by barrettj12 — COMMENTED** [review]
Very nice, just a few things that could maybe be elaborated.

File: worker/lease/doc.go [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+// Pin is a special case for making sure that a series upgrade doesn't change
+// the lease of an application. This means that the pinned lease will not
+// expire and the application will not lose its leadership. After the series
+// upgrade finishes it will be refreshed (extending its expiry date) and
+// finally unpinned, bringing operation back to normal.
```
Comment (tmihoc) [inline worker/lease/doc.go:21]: I'd try to roll this information into the list above, following Jordan's suggestion. Here it feels hard to parse because it's out of context. Also, it would be good to describe the other items in the list as well.

**Review by tmihoc — CHANGES_REQUESTED** [review]

File: worker/lease/doc.go [hunk 5]
```diff
@@ -0,0 +1,22 @@
+// Copyright 2023 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+// Package lease is a worker that manages leases, which are used by individual
```
[the remark this answers is not in the document]
Comment (barrettj12) [reply worker/lease/doc.go:4]: Or:
> Package lease defines a worker that ...

[the remark this answers is not in the document]
Comment (tmihoc) [reply worker/lease/doc.go:4]: I like Jordan's suggestion:

> Package lease defines the ... worker. This worker...

**Review by tmihoc — APPROVED** [review]
LGTM, thanks!
